### PR TITLE
Make location a dictionary and add it to SDFGState

### PR DESCRIPTION
**Pull request by definelicht**
Location doesn't seem to be used in many places, but might have external dependencies I don't know about.

### CODE REVIEW

**Comment by definelicht** [comment]
@tbennun This is one of those PRs that seems like a small thing, but has the potential to become a major shitshow 🙏

File: dace/data.py [hunk 1]
```diff
@@ -57,9 +57,9 @@ class Data(object):
                        default=dace.dtypes.StorageType.Default,
                        from_string=lambda x: dtypes.StorageType[x])
     location = Property(
-        dtype=str,  # Dict[str, symbolic]
```
Comment (tbennun) [inline dace/data.py]: we have the class `DictProperty` especially designed for this, with serialization support.

File: dace/data.py [hunk 2]
```diff
@@ -133,7 +133,7 @@ def __init__(self,
                  transient=False,
                  storage=dace.dtypes.StorageType.Default,
                  allow_conflicts=False,
-                 location='',
+                 location={},
```
Comment (tbennun) [inline dace/data.py]: That's generally not good in Python. It'll create one mutable object that if someone changes, will change for everyone who call this function. Please use `=None`  instead and `self.location = location or {}` inside the method.

File: dace/data.py [hunk 3]
```diff
@@ -268,7 +268,7 @@ def __init__(self,
                  transient=False,
                  allow_conflicts=False,
                  storage=dace.dtypes.StorageType.Default,
-                 location='',
+                 location={},
```
Comment (tbennun) [inline dace/data.py]: same

File: dace/data.py [hunk 4]
```diff
@@ -461,7 +461,7 @@ def __init__(self,
                  shape=None,
                  transient=False,
                  storage=dace.dtypes.StorageType.Default,
-                 location='',
+                 location={},
```
Comment (tbennun) [inline dace/data.py]: same

File: dace/graph/nodes.py [hunk 5]
```diff
@@ -257,15 +257,16 @@ class CodeNode(Node):
         denoted by an octagonal shape. """
 
     label = Property(dtype=str, desc="Name of the CodeNode")
-    location = Property(dtype=str,
-                        desc="CodeNode execution location descriptor",
-                        allow_none=True)
+    location = Property(
```
Comment (tbennun) [inline dace/graph/nodes.py]: `DictProperty`

File: dace/graph/nodes.py [hunk 6]
```diff
@@ -305,7 +306,7 @@ def __init__(self,
                  code_global="",
                  code_init="",
                  code_exit="",
-                 location="-1",
+                 location={},
```
Comment (tbennun) [inline dace/graph/nodes.py]: same

File: dace/graph/nodes.py [hunk 7]
```diff
@@ -257,15 +257,16 @@ class CodeNode(Node):
...
+        dtype=dict,
+        desc='Full storage location identifier (e.g., rank, GPU ID)',
+        default={})
     environments = SetProperty(
         str,
         desc="Environments required by CMake to build and run this code node.",
         default=set())
 
-    def __init__(self, label="", location=None, inputs=None, outputs=None):
+    def __init__(self, label="", location={}, inputs=None, outputs=None):
```
Comment (tbennun) [inline dace/graph/nodes.py]: same

File: dace/graph/nodes.py [hunk 8]
```diff
@@ -417,7 +418,7 @@ def __init__(self,
                  outputs: Set[str],
                  symbol_mapping: Dict[str, Any] = None,
                  schedule=dtypes.ScheduleType.Default,
-                 location="-1",
+                 location={},
```
Comment (tbennun) [inline dace/graph/nodes.py]: same

File: dace/libraries/blas/environments/cublas.py [hunk 9]
```diff
@@ -20,11 +20,11 @@ class cuBLAS:
     @staticmethod
     def handle_setup_code(node):
         location = node.location
-        if not location:
+        if not location or "cuda_device" not in node.location:
```
Comment (tbennun) [inline dace/libraries/blas/environments/cublas.py]: Somehow I prefer "gpu" for the GPU ID. Thoughts?

File: dace/libraries/blas/nodes/matmul.py [hunk 10]
```diff
@@ -406,7 +406,7 @@ class MatMul(dace.graph.nodes.LibraryNode):
     # Object fields
     dtype = dace.properties.TypeClassProperty(allow_none=True)
 
-    def __init__(self, name, dtype=None, location=None):
+    def __init__(self, name, dtype=None, location={}):
```
Comment (tbennun) [inline dace/libraries/blas/nodes/matmul.py]: same

File: dace/libraries/blas/nodes/transpose.py [hunk 11]
```diff
@@ -138,7 +138,7 @@ class Transpose(dace.graph.nodes.LibraryNode):
 
     dtype = dace.properties.TypeClassProperty(allow_none=True)
 
-    def __init__(self, name, dtype=None, location=None):
+    def __init__(self, name, dtype=None, location={}):
```
Comment (tbennun) [inline dace/libraries/blas/nodes/transpose.py]: same

File: dace/sdfg.py [hunk 12]
```diff
@@ -2464,7 +2470,12 @@ class SDFGState(OrderedMultiDiConnectorGraph, MemletTrackingView):
         desc="Measure execution statistics with given method",
         default=dtypes.InstrumentationType.No_Instrumentation)
 
-    def __init__(self, label=None, sdfg=None, debuginfo=None):
+    location = Property(
```
Comment (tbennun) [inline dace/sdfg.py]: `DictProperty`

File: dace/sdfg.py [hunk 13]
```diff
@@ -2861,17 +2878,17 @@ def add_access(self, array_or_stream_name: str,
...
+        self,
+        name: str,
+        inputs: Set[str],
+        outputs: Set[str],
+        code: str,
+        language: dtypes.Language = dtypes.Language.Python,
+        code_global: str = "",
+        code_init: str = "",
+        code_exit: str = "",
+        location: dict = {},
```
Comment (tbennun) [inline dace/sdfg.py]: same

File: dace/sdfg.py [hunk 14]
```diff
@@ -2891,16 +2908,16 @@ def add_tasklet(
...
-            debuginfo=None,
+        self,
+        sdfg: SDFG,
+        parent,
+        inputs: Set[str],
+        outputs: Set[str],
+        symbol_mapping: Dict[str, Any] = None,
+        name=None,
+        schedule=dtypes.ScheduleType.Default,
+        location={},
```
Comment (tbennun) [inline dace/sdfg.py]: same

File: dace/sdfg.py [hunk 15]
```diff
@@ -3044,7 +3061,7 @@ def add_mapped_tasklet(
             code_global="",
             code_init="",
             code_exit="",
-            location="-1",
+            location={},
```
Comment (tbennun) [inline dace/sdfg.py]: same

File: tests/wcr_cudatest.py [hunk 16]
```diff
@@ -19,7 +19,7 @@ def my_add_mapped_tasklet(
         code_global="",
         code_init="",
         code_exit="",
-        location="-1",
+        location={},
```
Comment (tbennun) [inline tests/wcr_cudatest.py]: same

File: dace/data.py (hunk 1 above)
Comment (definelicht) [reply dace/data.py]: How does this work for default? We need to set this now because of serialization, right?

Comment (definelicht) [reply dace/data.py]: `DictProperty` has `value_type` as mandatory argument, but location does not enforce this. Should I change `DictProperty` so this is optional?

File: dace/data.py [hunk 17]
```diff
@@ -71,7 +71,7 @@ def __init__(self, dtype, shape, transient, storage, location, toplevel,
         self.shape = shape
         self.transient = transient
         self.storage = storage
-        self.location = location
+        self.location = location if location is not None else {}
```
Comment (tbennun) [inline dace/data.py]:
```suggestion
        self.location = location or {}
```

File: tests/wcr_cudatest.py (hunk 16 above)
Comment (tbennun) [inline tests/wcr_cudatest.py]: oops

**Pull request merged**